src/simulators/unitary/unitarymatrix.hpp

```diff
@@ -224,6 +252,63 @@ void UnitaryMatrix<data_t>::set_num_qubits(size_t num_qubits) {
   BaseVector::set_num_qubits(2 * num_qubits);
 }
 
+template <class data_t>
+complex_t UnitaryMatrix<data_t>::trace() const {
+  const int_t NROWS = rows_;
+  const int_t DIAG = NROWS + 1;
+  double val_re = 0.;
+  double val_im = 0.;
+#pragma omp parallel reduction(+:val_re, val_im) if (BaseVector::num_qubits_ > BaseVector::omp_threshold_ && BaseVector::omp_threads_ > 1) num_threads(BaseVector::omp_threads_)
+  {
+#pragma omp for
+  for (int_t k = 0; k < NROWS; ++k) {
+    val_re += std::real(BaseVector::data_[k * DIAG]);
+    val_im += std::imag(BaseVector::data_[k * DIAG]);
+  }
+  }
+  return complex_t(val_re, val_im);
+}
+
+
+//------------------------------------------------------------------------------
+// Check Identity
+//------------------------------------------------------------------------------
+
+template <class data_t>
+std::pair<bool, double> UnitaryMatrix<data_t>::check_identity() const {
+  // To check if identity we first check we check that:
+  // 1. U(0, 0) = exp(i * theta)
+  // 2. U(i, i) = U(0, 0)
+  // 3. U(i, j) = 0 for j != i 
+  auto failed = std::make_pair(false, 0.0);
```

Comment:
i feel like this variable should be called `passed` instead of `failed`, because when passing, you set it to `true`.

Reply:
This variable is called `failed` because it's storing the return value for when it fails to be unitary ie: `return failed;` when it fails. A separate pair containing the phase is constructed and returned when it doesn't fail.